src/timeseriesflattener/spec_processors/temporal.py

```diff
@@ -215,50 +215,72 @@ def _create_step_frames(
+
+        result_frames = list()
+        for step in range(len(step_date_ranges) - 1):
+            step_predictiontime_frame, step_value_frame = _create_step_frames(
+                predictiontime_frame, spec, step_date_ranges, step
+            )
 
-    result_frames = list()
-    for step in range(len(step_date_ranges) - 1):
-        step_predictiontime_frame, step_value_frame = _create_step_frames(
-            predictiontime_frame, spec, step_date_ranges, step
-        )
+            step_aggregated_value_frames = _process_temporal_spec(
+                spec, step_predictiontime_frame, step_value_frame
+            )
 
-        step_aggregated_value_frames = (
-            Iter(spec.normalised_lookperiod)
-            .map(
-                lambda lookperiod: _slice_and_aggregate_spec(
-                    timedelta_frame=_get_timedelta_frame(
-                        predictiontime_frame=step_predictiontime_frame, value_frame=step_value_frame
-                    ),
-                    masked_aggregator=lambda sliced_frame: _aggregate_masked_frame(
-                        aggregators=spec.aggregators,
-                        fallback=spec.fallback,
-                        masked_frame=sliced_frame,
-                    ),
-                    time_masker=lambda timedelta_frame: _mask_outside_lookperiod(
-                        timedelta_frame=timedelta_frame,
-                        lookperiod=lookperiod,
-                        column_prefix=spec.column_prefix,
-                        value_col_names=spec.value_frame.value_col_names,
-                    ),
+            result_frames += [
+                horizontally_concatenate_dfs(
+                    step_aggregated_value_frames,
+                    pred_time_uuid_col_name=step_predictiontime_frame.pred_time_uuid_col_name,
                 )
-            )
-            .flatten()
-            .to_list()
+            ]
+
+        return ProcessedFrame(
+            df=pl.concat(result_frames, how="vertical"),
+            pred_time_uuid_col_name=predictiontime_frame.pred_time_uuid_col_name,
         )
 
-        result_frames += [
-            horizontally_concatenate_dfs(
-                step_aggregated_value_frames,
-                pred_time_uuid_col_name=step_predictiontime_frame.pred_time_uuid_col_name,
-            )
-        ]
+    else:
+        aggregated_value_frames = _process_temporal_spec(
+            spec, predictiontime_frame, spec.value_frame
+        )
 
-    return ProcessedFrame(
-        df=pl.concat(result_frames, how="vertical"),
-        pred_time_uuid_col_name=predictiontime_frame.pred_time_uuid_col_name,
-    )
+        return ProcessedFrame(
+            df=horizontally_concatenate_dfs(
+                dfs=aggregated_value_frames,
+                pred_time_uuid_col_name=predictiontime_frame.pred_time_uuid_col_name,
+            ),
+            pred_time_uuid_col_name=predictiontime_frame.pred_time_uuid_col_name,
+        )
```

Comment:
Duplicate functionality here. Challenging you a bit; I'm sure you can refactor it 😉